test/unit/core/bit_manipulation_test.cpp

```diff
@@ -81,29 +83,135 @@ class unsigned_operations : public ::testing::Test
+                                                         << position;
+        }
+    }
+}
+
+TYPED_TEST(unsigned_operations, count_leading_zeros)
+{
+    using unsigned_t = TypeParam;
+    constexpr size_t t1 = count_leading_zeros<unsigned_t>(0b0001);
+    constexpr size_t t2 = count_leading_zeros<unsigned_t>(0b0101);
+    constexpr size_t t3 = count_leading_zeros<unsigned_t>(0b0010);
+    constexpr size_t t4 = count_leading_zeros<unsigned_t>(0b0110);
+    constexpr size_t t5 = count_leading_zeros<unsigned_t>(0b0100);
+    constexpr size_t t6 = count_leading_zeros<unsigned_t>(0b10100000);
+    EXPECT_EQ(t1, sizeof_bits<unsigned_t> - 1u);
+    EXPECT_EQ(t2, sizeof_bits<unsigned_t> - 3u);
+    EXPECT_EQ(t3, sizeof_bits<unsigned_t> - 2u);
+    EXPECT_EQ(t4, sizeof_bits<unsigned_t> - 3u);
+    EXPECT_EQ(t5, sizeof_bits<unsigned_t> - 3u);
+    EXPECT_EQ(t6, sizeof_bits<unsigned_t> - 8u);
+
+    for (uint8_t cnt = 0; cnt < sizeof_bits<unsigned_t>; ++cnt)
+    {
+        unsigned_t start = std::numeric_limits<unsigned_t>::max() >> cnt;
+        unsigned_t end = start >> 1u;
+        for (unsigned_t n = start, k = 0u; n < end && k < max_iterations; ++n, ++k)
+        {
+            EXPECT_EQ(sizeof_bits<unsigned_t> - sdsl::bits::hi(n) - 1, n) << "[SDSL] n " << n << " should have "
+                                                                          << cnt << " leading zeros.";
+            EXPECT_EQ(count_leading_zeros(n), cnt) << "n " << n << " should have " << cnt << " leading zeros.";
+        }
+    }
+}
+
+TYPED_TEST(unsigned_operations, count_trailing_zeros)
+{
+    using unsigned_t = TypeParam;
+    constexpr size_t zero  = count_trailing_zeros<unsigned_t>(0b0001);
+    constexpr size_t zero2 = count_trailing_zeros<unsigned_t>(0b0101);
+    constexpr size_t one1  = count_trailing_zeros<unsigned_t>(0b0010);
+    constexpr size_t one2  = count_trailing_zeros<unsigned_t>(0b0110);
+    constexpr size_t two   = count_trailing_zeros<unsigned_t>(0b0100);
+    constexpr size_t five  = count_trailing_zeros<unsigned_t>(0b10100000);
+    EXPECT_EQ(zero,  0u);
+    EXPECT_EQ(zero2, 0u);
+    EXPECT_EQ(one1,  1u);
+    EXPECT_EQ(one2,  1u);
+    EXPECT_EQ(two,   2u);
+    EXPECT_EQ(five,  5u);
+
+    for (uint8_t cnt = 0; cnt < sizeof_bits<unsigned_t>; ++cnt)
+    {
+        unsigned_t start = std::numeric_limits<unsigned_t>::max() << cnt;
+        unsigned_t end = start << 1u;
+        for (unsigned_t n = start, k = 0u; n < end && k < max_iterations; ++n, ++k)
+        {
+            EXPECT_EQ(sdsl::bits::lo(n), cnt) << "[SDSL] n " << n << " should have " << cnt << " trailing zeros.";
+            EXPECT_EQ(count_trailing_zeros(n), cnt) << "n " << n << " should have " << cnt << " trailing zeros.";
+        }
+    }
+}
+
+// http://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
+template <std::UnsignedIntegral unsigned_t>
+unsigned_t permute_bits(unsigned_t v)
```

Comment:
and we need this only for the tests?

Reply:
Yes, to follow the scheme of the other tests, I wanted to compute all integers with a given number of 1's in them.